### PR TITLE
Database Decodable (And More!)

### PR DESCRIPTION
**Pull request by stephencelis**
Adds a bunch of fun things around our data serialization/deserialization!

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -48,13 +48,13 @@ route-partial-iso:
 		--output ./Sources/PointFree/__Generated__/DerivedPartialIsos.swift
 
 postgres-mm:
-	-@mkdir -p "$(POSTGRES_PATH)"
-	-@echo "$$POSTGRES_MODULE_MAP" > "$(POSTGRES_PATH)/module.map"
-	-@echo "$$POSTGRES_SHIM_H" > "$(POSTGRES_PATH)/shim.h"
+	-@sudo mkdir -p "$(POSTGRES_PATH)"
+	-@echo "$$POSTGRES_MODULE_MAP" | sudo tee "$(POSTGRES_PATH)/module.map" > /dev/null
```
Comment (stephencelis) [inline Makefile]: Quick module map fix for playgrounds!

File: Sources/PointFree/About.swift [hunk 2]
```diff
@@ -5,21 +5,22 @@ import HttpPipeline
 import HttpPipelineHtmlSupport
 import Prelude
 import Styleguide
+import Tuple
 
 let aboutResponse: Middleware<StatusLineOpen, ResponseEnded, Prelude.Unit, Data> =
-  requestContextMiddleware
+  _requestContextMiddleware
```
Comment (stephencelis) [inline Sources/PointFree/About.swift]: This snuck in from our sesh last week. Should I remove this part of the diff?

File: Sources/PointFree/Database.swift [hunk 3]
```diff
@@ -16,19 +16,23 @@ public struct Database {
...
 
-  public struct User {
-    let email: String
-    let gitHubUserId: Int
-    let gitHubAccessToken: String
-    let id: UUID
-    let name: String
-    let subscriptionId: UUID?
+  public struct User: Decodable {
+    public let email: EmailAddress
```
Comment (stephencelis) [inline Sources/PointFree/Database.swift]: We now ensure that, if we load the user, the email is its own new type.

File: Sources/PointFree/Database.swift [hunk 4]
```diff
@@ -16,19 +16,23 @@ public struct Database {
...
+    public let gitHubUserId: GitHub.User.Id
```
Comment (stephencelis) [inline Sources/PointFree/Database.swift]: The GitHub user id is also its own type!

File: Sources/PointFree/Database.swift [hunk 5]
```diff
@@ -16,19 +16,23 @@ public struct Database {
...
+    public let gitHubAccessToken: String
```
Comment (stephencelis) [inline Sources/PointFree/Database.swift]: We could type this, too? 🤔

File: Sources/PointFree/Database.swift [hunk 6]
```diff
@@ -82,27 +86,15 @@ private func createUser(with envelope: GitHub.UserEnvelope) -> EitherIO<Error, P
 }
 
 private func fetchUser(with token: GitHub.AccessToken) -> EitherIO<Error, Database.User?> {
-  return execute(
+  return firstRow(
```
Comment (stephencelis) [inline Sources/PointFree/Database.swift]: A helper that's basically like Active Record's `first` using `Decodable` under the hood.

File: Sources/PointFree/Database.swift [hunk 7]
```diff
@@ -82,27 +86,15 @@ private func createUser(with envelope: GitHub.UserEnvelope) -> EitherIO<Error, P
...
     """
     SELECT "email", "github_user_id", "github_access_token", "id", "name"
     FROM "users"
     WHERE "github_access_token" = $1
     LIMIT 1
     """,
     [token.accessToken]
-    )
-    .map { result -> Database.User? in
-      let uuid = result["id"]?.array?.first?.wrapped.string.flatMap(UUID.init(uuidString:))
```
Comment (stephencelis) [inline Sources/PointFree/Database.swift]: Nice to get rid of this messiness!

File: Sources/PointFree/DatabaseCodable.swift [hunk 8]
```diff
@@ -0,0 +1,397 @@
+import Foundation
+import PostgreSQL
+import Prelude
+
+final class DatabaseDecoder: Decoder {
```
Comment (stephencelis) [inline Sources/PointFree/DatabaseCodable.swift]: This took awhile to get working, but hopefully makes our lives much easier!

File: Sources/PointFree/Routers/PointFreeRoutes.swift [hunk 9]
```diff
@@ -10,12 +10,12 @@ public enum Route: DerivePartialIsos {
   case episodes(tag: Tag?)
   case gitHubCallback(code: String?, redirect: String?)
   case home(signedUpSuccessfully: Bool?)
-  case launchSignup(email: String)
+  case launchSignup(EmailAddress)
```
Comment (stephencelis) [inline Sources/PointFree/Routers/PointFreeRoutes.swift]: Nice to encode meaning in the types directly!

File: Sources/PointFree/Routers/PointFreeRoutes.swift [hunk 10]
```diff
@@ -39,7 +39,7 @@ private let routers: [Router<Route>] = [
     <¢> get %> queryParam("success", opt(.bool)) <% end,
 
   Route.iso.launchSignup
-    <¢> post %> formField("email") <% lit("launch-signup") <% end,
+    <¢> post %> formField("email", .rawRepresentable) <% lit("launch-signup") <% end,
```
Comment (stephencelis) [inline Sources/PointFree/Routers/PointFreeRoutes.swift]: `RawRepresentable` makes it easy to decode tagged types in our router!

File: Sources/PointFree/Stripe.swift [hunk 11]
```diff
@@ -24,30 +24,27 @@ public struct Stripe {
     fetchSubscription: PointFree.fetchSubscription
   )
 
-  public struct Cents: SingleValueCodable {
-    public let rawValue: Int
-    public init(rawValue: Int) {
-      self.rawValue = rawValue
-    }
-  }
+  public typealias Cents = Tagged<Stripe, Int>
```
Comment (stephencelis) [inline Sources/PointFree/Stripe.swift]: Fun using `Tagged` everywhere for almost free type-safety!

File: Sources/PointFree/Stripe.swift [hunk 12]
```diff
@@ -124,32 +123,40 @@ public struct Stripe {
...
+  return stripeDataTask("https://api.stripe.com/v1/subscriptions/\(id.rawValue)", .delete)
 }
 
-private func createCustomer(token: String) -> EitherIO<Prelude.Unit, Stripe.Customer> {
-  return stripeDataTask("https://api.stripe.com/v1/customers", .post(["source": token]))
+private func createCustomer(user: Database.User, token: Stripe.Token.Id)
+  -> EitherIO<Prelude.Unit, Stripe.Customer> {
+
+    return stripeDataTask("https://api.stripe.com/v1/customers", .post([
+      "description": user.id.unwrap.uuidString,
```
Comment (stephencelis) [inline Sources/PointFree/Stripe.swift]: Getting codability around form data will be nice.

File: Sources/PointFree/Stripe.swift [hunk 13]
```diff
@@ -207,19 +214,6 @@ private func attachStripeAuthorization(_ headers: [String: String]?) -> [String:
     |> key("Authorization") .~ ("Basic " + secret) // TODO: Use key path subscript
 }
 
-public protocol SingleValueCodable: Codable, RawRepresentable {}
-
-extension SingleValueCodable where RawValue: Codable {
```
Comment (stephencelis) [inline Sources/PointFree/Stripe.swift]: Baking this functionality into `Tagged` instead.

File: Sources/PointFree/Subscribe.swift [hunk 14]
```diff
@@ -1,20 +1,22 @@
...
 import Prelude
+@testable import Tuple
 
-typealias SubscribeData = (plan: Stripe.Plan.Id, token: String)
+typealias SubscribeData = (plan: Stripe.Plan.Id, token: Stripe.Token.Id)
 
-let subscribeResponse: Middleware<StatusLineOpen, ResponseEnded, SubscribeData, Data> =
-  requestContextMiddleware
+let subscribeResponse =
+  map { $0 .*. unit }
```
Comment (stephencelis) [inline Sources/PointFree/Subscribe.swift]: This lifts a singleton value into a `T2`.

File: Makefile (hunk 1 above)
Comment (mbrandonw) [reply Makefile]: nice! ill do the same for swift-web!

Comment (stephencelis) [reply Makefile]: I think I already did? 🤔

**Pull request merged**